syntax = "proto3";

package bookstore;

import "google/protobuf/empty.proto";

service BooksDatabase {
    rpc Read(ReadRequest) returns (ReadResponse);
    rpc Write(WriteRequest) returns (WriteResponse);
    rpc Update(WriteRequest) returns (WriteResponse);
    rpc Adjust(AdjustRequest) returns (WriteResponse);
    rpc Exists(ReadRequest) returns (ExistsResponse);
    // 2PC methods
    rpc PrepareAdjust(PrepareAdjustRequest) returns (PrepareAdjustResponse);
    rpc FinalizeAdjust(FinalizeAdjustRequest) returns (google.protobuf.Empty);
}

message ReadRequest {
    string key = 1;
}

message ReadResponse {
    int32 value = 1;
}

message WriteRequest {
    string key = 1;
    int32 value = 2;
}

message WriteResponse {
    bool success = 1;
}

message AdjustRequest {
    string key = 1;
    int32 amount = 2;
    bool allowNegativeResult = 3;
}

message ExistsResponse {
    bool exists = 1;
}

message PrepareAdjustRequest {
    string id = 1;
    AdjustRequest request = 2;
}

message PrepareAdjustResponse {
    bool ready = 1;
}

message FinalizeAdjustRequest {
    string id = 1;
    bool abort = 2;
}
